# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [6]:
file="data/step3_output.csv"
df=pd.read_csv(file)
df = df.drop(df.columns[0], axis=1)
df.head()
print(df.shape)


(253, 26)


## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [7]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)

In [8]:
print(df.shape)
df.head(10)

(253, 60)


Name  summit_elev  vertical_drop  trams  \
0            Eaglecrest Ski Area         2600           1540      0   
1               Hilltop Ski Area         2090            294      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4         Boreal Mountain Resort         7700            500      0   
5                    Dodge Ridge         8200           1600      0   
6               Donner Ski Ranch         8012            750      0   
7                  June Mountain        10090           2590      0   
8                       Kirkwood         9800           2000      0   
9            Mt. Shasta Ski Park         6890           1435      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       0       4  ...             0   
1        0.0          0          0     0       1       0  ...             0   
2        0.0          0          1     2       3       1  ...             0   
3        0.0          0          0     0       1       3  ...             0   
4        0.0          0          1     1       3       1  ...             0   
5        0.0          0          0     1       2       5  ...             0   
6        0.0          0          0     0       1       5  ...             0   
7        0.0          0          2     0       0       4  ...             0   
8        0.0          0          2     1       5       1  ...             0   
9        0.0          0          0     0       3       0  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   
5             0          0     0        0         0           0   
6             0          0     0        0         0           0   
7             0          0     0        0         0           0   
8             0          0     0        0         0           0   
9             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  
5              0          0        0  
6              0          0        0  
7              0          0        0  
8              0          0        0  
9              0          0        0  

[10 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:

#First we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)


**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable 
#with parameter X_test

y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [14]:
explained_variance_score(y_test, y_pred)

-2.9125582402820147e+22

In [15]:
mean_absolute_error(y_test, y_pred)

455384110291.57965

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [16]:
lm.intercept_

115195980525.66452

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [17]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient',ascending=False).head(10)

#Here, the results diverge significantly from the expected ones...Why does including states make the model
#so inaccurate?

Coefficient
fastEight     3.159296e+13
total_chairs  2.791570e+13
trams         2.084898e+13
fastSixes     1.996033e+13
surface       1.513332e+13
double        1.486920e+13
triple        1.373280e+13
quad          8.623470e+12
fastQuads     7.957927e+12
Michigan      4.424835e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [18]:
#Read file
file='data/step3_output.csv'
df=pd.read_csv(file)

#Drop columns and extract AdultWeekend
X=df.drop([df.columns[0],'AdultWeekend','Name','state'], axis=1)
y=df.AdultWeekend

#Standardizing/Scaling
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)

#Splitting
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

#Fitting
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

#Evaluating
y_pred = model.predict(X_test)
print("Explained Variance Score: ", explained_variance_score(y_test, y_pred))
print("Mean Absolute Error: ", mean_absolute_error(y_test, y_pred))

Explained Variance Score:  0.7133091719139912
Mean Absolute Error:  6.420847162967177


In [19]:
lm.intercept_

58.138349165083014

In [20]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient',ascending=False).head(10)

Coefficient
AdultWeekday         12.212222
SkiableTerrain_ac     2.728044
vertical_drop         2.722381
quad                  2.055216
daysOpenLastYear      1.552390
clusters              1.504547
surface               1.240419
Runs                  1.203194
projectedDaysOpen     1.062290
fastQuads             1.006078

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [21]:
#base_elev had already been removed in Step 3.

#Read file
file="data/step3_output.csv"
df=pd.read_csv(file)

#Drop columns and extract AdultWeekend
X=df.drop([df.columns[0],'AdultWeekend','Name','state','summit_elev'], axis=1)
y=df.AdultWeekend

#Standardizing/Scaling
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)

#Splitting
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

#Fitting
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

#Evaluating
y_pred = model.predict(X_test)
print("Explained Variance Score: ", explained_variance_score(y_test, y_pred))
print("Mean Absolute Error: ", mean_absolute_error(y_test, y_pred))

Explained Variance Score:  0.7138913961621356
Mean Absolute Error:  6.411371722477571


In [22]:
lm.intercept_

58.14180869231809

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.913e+22 | 455384110291.580 |-|
| Model 2. | 0.713 | 6.421 |'state'|
| Model 3. | 0.714 | 6.411 |'state','summit_elev'|

Model Selection: Model 3.